[Amazon CAPI] - Changing how custom attribute field works

joe-ayoub-segment opened the pull request:
This is a breaking change for customers who have the Amazon CAPI Destination installed ( for private beta testing )

Custom Attributes field is being converted into a regular object field, defaulting to key:value UI display. 

Types for each attribute will now be inferred by Segment before being sent to Amazon. 
null and undefined - these will be dropped
number, boolean, object - these will be stringified and sent
string - these will be sent

Customers already using the Integration will need to reconfigure the Custom Attributes field.  

## Testing

Tested locally

bansrav commented on packages/destination-actions/src/destinations/amazon-conversions-api/trackConversion/utils.ts:
```diff
@@ -363,6 +366,28 @@ export function prepareEventData(payload: Payload, settings: Settings): EventDat
+    if (value === undefined || value === null) return
+    let dataType: 'STRING' | 'NUMBER' | 'BOOLEAN' = 'STRING'
+
+    if (typeof value === 'number') {
```
can we simplify this logic based on the below context. Just pass everything as string value. Ideally, the dataType should be hidden in our CAPI spec.

```
I checked the CAPI v2 code on our side and discussed about the usage of datatype with our product manager. In code, we never use datatype field. We only ingest name and value attributes from custom attributes. Our product manager also confirmed that we can safely remove the datatype. Also, only brand, category, productId and attr1 - attr10 custom attributes are used for reporting
```

bansrav commented on packages/destination-actions/src/destinations/amazon-conversions-api/trackConversion/__tests__/utils.test.ts:
```diff
@@ -590,23 +590,80 @@ describe('trackConversion utils', () => {
-          { name: 'color', dataType: 'STRING', value: 'blue' },
-          { name: 'size', dataType: 'STRING', value: 'medium' }
-        ],
+        customAttributes: {
```
this is not a real world use0case for CAPI. Only brand, category, productId and attr1 to attr10 custom attributes are supported and used for reporting therefore all numbers, booleans or attributes not falling into the this criteria are simply ignored.

bansrav commented on packages/destination-actions/src/destinations/amazon-conversions-api/trackConversion/generated-types.ts:
```diff
@@ -119,19 +119,8 @@ export interface Payload {
    * Custom attributes associated with the event to provide additional context.
    */
   customAttributes?: {
```
can we add description here similar to what I have in here - https://github.com/segmentio/action-destinations/pull/3052/files

It's important to emphasize on the attributes being supported

`description: 'Custom attributes associated with the event to provide additional context. Only brand, category, productId and attr1 - attr10 custom attributes are used for reporting.'`

joe-ayoub-segment commented:
hi @bansrav I incorporated your comments. Please take a look. I also removed the 2 fields you removed in your [PR](https://github.com/segmentio/action-destinations/pull/3052).

bansrav commented on packages/destination-actions/src/destinations/amazon-conversions-api/trackConversion/utils.ts:
```diff
@@ -363,6 +366,24 @@ export function prepareEventData(payload: Payload, settings: Settings): EventDat
+    if (value === undefined || value === null) return
+
+    if (!CustomAttributeV1Names.includes(key as CustomAttributeV1Name)) {
+      throw new PayloadValidationError(
```
This won't work. We want to ingest all key/value pairs from clients as there will be use-cases coming later this year related to audience creation from these custom attributes for CAPI, but only brand, category, productId, attr1-attr-10 are used for reporting. So, here is what we should do

1. Display brand, category, productId and attr1-attr-10 explicitly in Segment UI under custom Attributes but also provide option to customer to add more key/value pairs (whatever key name they want) if they like. 
2. Send all those custom attributes to CAPI v2.

joe-ayoub-segment replied:
Thanks for the feedback @bansrav - I've updated the PR as per my understanding of your requirements.
